versioned_docs/version-2022sp/lecture7.md

```diff
@@ -0,0 +1,518 @@
+```
+
+`setDoc` will write to the document `jason` within the `people` collection, creating it
+if it does not yet exist.
+
+An alternative approach if you strictly want to add and never overwrite a doc is `addDoc`.
+Here's the same code but using `addDoc`.
+
+```typescript
+await addDoc(peopleCollectionRef, { year: '2023', notes_writer: true });
+```
+
+Note, though, that using the latter means the `doc.id` generated will be random, so
+you would probably want to add the name into a field. The same behavior with autogenerated ids
+can be replicated with setDoc by omitting the name of the document in the `doc()` call.
+
+### Read
+
+To read data in Firestore, we first need to take a snapshot of the data we want to read. Then,
+we check if that snapshot of the document we want actually exists before trying to extract its data.
+
+Here's a simple example logging my data. Notice I log `docSnap.data()` and not `docSnap` because
+the former is where the data actually resides.
+
+```typescript
+const docSnap = await getDoc(jasonDocRef);
+if (docSnap.exists()) {
+  console.log('Document data:', docSnap.data());
+} else {
+  console.log('No such document!');
+}
+```
+
+We can also perform querying on collections, where we filter the database with certain
+criteria. Here's a simple example of looking for all people in `people` who are in the year 2023.
+
+```typescript
+const query = query(peopleCollectionRef, where('year', '==', 2023));
+const querySnapshot = await getDocs(query);
+querySnapshot.forEach((doc) => {
+  console.log(doc.id, ' => ', doc.data());
+});
+```
+
+Line by line, we first construct a query for the collection that narrows it down for us.
+Next, we take a snapshot of all such documents in the query, and lastly, we perform the same
+`doc.data()` approach. We do not need to check if each doc exists because we are using a for loop,
+and if a doc does not exist, it simply would not be part of the for loop.
+
+We can also order our search results. Here is an example of ordering by alphabetical order descending.
+
+```typescript
+const query = query(peopleCollectionRef, orderBy('name', desc));
+// using getDocs to get all documents that match
+const querySnapshot = await getDocs(query);
+querySnapshot.forEach((doc) => {
+  console.log(doc.id, ' => ', doc.data());
+});
+```
+
+We can actually combine multiple `queryConstraints` (like `where` and `orderBy`)
+by listing them out as multiple parameters within `query`.
+
+### Update
+
+Updating a document will only replace the specified fields within a doc and maintain
+unmodified fields. So the following code keep the `notes_writer` field but change `year` and
+`cat_or_dog`.
+
+```typescript
+await updateDoc(jasonDocRef, { year: '2022', cat_or_dog: 'cat' });
+```
+
+### Delete
+
+Deleting a document removes it from the collection.
+
+```typescript
+await deleteDoc(jasonDocRef);
+```
+
+## Sample code
+
+```typescript title="index.ts"
```

Comment:
This code block is super long and uses firebase-admin, so it's not a good example for using the frontend firebase lib (esp v9)